File: packages/os-domain/src/access-domain/control.ts
import type { PrincipalId } from "../types.js";
import type { AccessDomainForest } from "./forest.js";
import { requireDomain } from "./forest.js";
import type { AccessRealm } from "./realm.js";
import {
  admitsSharing,
  assertAdmitsSharing,
  refuseAccessDomain,
} from "./realm.js";

/**
 * What a principal may do in a domain.
 *
 * The same three rungs as organization and ADR 0038 project memberships,
 * deliberately not a fourth vocabulary.
 */
export type DomainRole = "member" | "admin" | "owner";

/** Weakest rung first, so an index comparison *is* the ladder. */
export const DOMAIN_ROLE_LADDER = ["member", "admin", "owner"] as const;

/** How far an assignment reaches. */
export type ControlScope = "domain_only" | "subtree";

function rung(role: DomainRole): number {
  return DOMAIN_ROLE_LADDER.indexOf(role);
}

/** Whether `role` covers `wanted`. Owning implies administering, and so on. */
export function roleCovers(role: DomainRole, wanted: DomainRole): boolean {
  return rung(role) >= rung(wanted);
}

/**
 * Whether this role may create, rename, move and retire domains, and hand
 * control to others.
 */
export function canAdminister(role: DomainRole): boolean {
  return role === "owner" || role === "admin";
}

/** Whether this role may hand out ownership. Only an owner may. */
export function canTransferOwnership(role: DomainRole): boolean {
  return role === "owner";
}

/** One grant of control. */
export interface ControlAssignment {
  /**
   * The handle an operator revokes. An assignment without its own id can be
   * withdrawn only by describing it, which is how the wrong one gets withdrawn.
   */
  readonly id: string;
  readonly domainId: string;
  readonly principalId: PrincipalId;
  readonly role: DomainRole;
  readonly scope: ControlScope;
}

/**
 * The answer to "what may this principal do here", and where it came from.
 *
 * `viaDomainId` is the domain the winning assignment was made at, which is what
 * an operator needs in order to revoke it.
 */
export interface EffectiveControl {
  readonly domainId: string;
  readonly principalId: PrincipalId;
  readonly role: DomainRole;
  readonly viaDomainId: string;
}

/** Every control assignment in a realm, keyed by assignment id. */
export interface ControlIndex {
  readonly assignments: ReadonlyMap<string, ControlAssignment>;
}

export function emptyControlIndex(): ControlIndex {
  return { assignments: new Map() };
}

export function controlAt(
  index: ControlIndex,
  domainId: string,
): ControlAssignment[] {
  return [...index.assignments.values()].filter(
    (assignment) => assignment.domainId === domainId,
  );
}

/**
 * Record an assignment made by the platform or the realm's owner.
 *
 * The realm comes from the forest rather than a separate argument, so there is
 * no call shape that checks one realm's sharing rule while writing into
 * another realm's estate.
 */
export function insertControl(
  index: ControlIndex,
  forest: AccessDomainForest,
  assignment: ControlAssignment,
): ControlIndex {
  requireDomain(forest, assignment.domainId);
  if (index.assignments.has(assignment.id)) {
    throw refuseAccessDomain(
      "conflict",
      `Control assignment ${assignment.id} already exists`,
      { id: assignment.id },
    );
  }
  const held = controlAt(index, assignment.domainId).some(
    (existing) => existing.principalId === assignment.principalId,
  );
  if (held) {
    throw refuseAccessDomain(
      "conflict",
      `${assignment.principalId} already holds control at ${assignment.domainId}`,
      { principalId: assignment.principalId, domainId: assignment.domainId },
    );
  }
  assertRealmAdmits(index, forest.realm, assignment.principalId);
  const assignments = new Map(index.assignments);
  assignments.set(assignment.id, assignment);
  return { assignments };
}

/**
 * Record an assignment one principal is handing to another.
 *
 * Delegation never widens: without this check an admin could mint an owner
 * below themselves and take the estate.
 */
export function insertDelegatedControl(
  index: ControlIndex,
  forest: AccessDomainForest,
  grantor: PrincipalId,
  assignment: ControlAssignment,
): ControlIndex {
  const held = effectiveControl(index, forest, assignment.domainId, grantor);
  if (held === undefined) {
    throw refuseAccessDomain(
      "control_widen",
      `${grantor} holds no control at ${assignment.domainId}`,
      { grantor, domainId: assignment.domainId },
    );
  }
  if (!canAdminister(held.role)) {
    throw refuseAccessDomain(
      "control_widen",
      `${grantor} is only a ${held.role} at ${assignment.domainId}`,
      { grantor, role: held.role },
    );
  }
  if (!roleCovers(held.role, assignment.role)) {
    throw refuseAccessDomain(
      "control_widen",
      `${grantor} holds ${held.role} and cannot grant ${assignment.role}`,
      { grantor, held: held.role, granting: assignment.role },
    );
  }
  if (
    canTransferOwnership(assignment.role) &&
    !canTransferOwnership(held.role)
  ) {
    throw refuseAccessDomain(
      "control_widen",
      `${grantor} cannot hand out ownership of ${assignment.domainId}`,
      { grantor, domainId: assignment.domainId },
    );
  }
  return insertControl(index, forest, assignment);
}

/** Withdraw an assignment. */
export function removeControl(index: ControlIndex, id: string): ControlIndex {
  const assignments = new Map(index.assignments);
  assignments.delete(id);
  return { assignments };
}

/**
 * What `principalId` may do at `domainId`, counting inherited control.
 *
 * Walks from the domain towards its root, taking the strongest role that
 * applies: assignments at the domain itself whatever their scope, and `subtree`
 * assignments from ancestors. The walk stops at an `isolated` domain, which is
 * what isolation means.
 */
export function effectiveControl(
  index: ControlIndex,
  forest: AccessDomainForest,
  domainId: string,
  principalId: PrincipalId,
): EffectiveControl | undefined {
  let best: EffectiveControl | undefined;
  const seen = new Set<string>();
  let cursor: string | undefined = domainId;
  let inherited = false;
  while (cursor !== undefined) {
    if (seen.has(cursor)) {
      throw refuseAccessDomain(
        "cycle",
        `Control walk from ${domainId} revisits ${cursor}`,
        { from: domainId, at: cursor },
      );
    }
    seen.add(cursor);
    const node = requireDomain(forest, cursor);
    for (const assignment of controlAt(index, cursor)) {
      if (assignment.principalId !== principalId) continue;
      if (inherited && assignment.scope !== "subtree") continue;
      if (best === undefined || rung(assignment.role) > rung(best.role)) {
        best = {
          domainId,
          principalId,
          role: assignment.role,
          viaDomainId: cursor,
        };
      }
    }
    if (node.inheritance === "isolated") break;
    cursor = node.parentId;
    inherited = true;
  }
  return best;
}

/** Refuse an administrative action by a principal who cannot administer. */
export function assertMayAdminister(
  index: ControlIndex,
  forest: AccessDomainForest,
  domainId: string,
  principalId: PrincipalId,
): EffectiveControl {
  const held = effectiveControl(index, forest, domainId, principalId);
  if (held !== undefined && canAdminister(held.role)) return held;
  throw refuseAccessDomain(
    "control_widen",
    `${principalId} cannot administer ${domainId}`,
    { principalId, domainId },
  );
}

/**
 * Refuse a second principal in a realm that does not share.
 *
 * ADR 0038's personal project refuses membership outright, so nesting domains
 * inside one cannot become a back door to sharing it.
 */
function assertRealmAdmits(
  index: ControlIndex,
  realm: AccessRealm,
  principalId: PrincipalId,
): void {
  if (admitsSharing(realm)) return;
  const stranger = [...index.assignments.values()].some(
    (existing) => existing.principalId !== principalId,
  );
  if (stranger) assertAdmitsSharing(realm, "a second controlling principal");
}

File: crates/domain/src/access_domain/control.rs
//! Who administers an access domain, and how control inherits down a subtree.
//!
//! Control is **not** data-plane permission. Holding [`DomainRole::Owner`] on a
//! domain lets a principal rearrange domains and issue beneath them; it does
//! not unwrap vault ciphertext or exercise a connection. Those remain a
//! Grant+Intent decision, so nothing here is a second authority model.
//!
//! Three rules shape the resolution in [`ControlIndex::effective_control`]:
//!
//! 1. **Inheritance flows down, never up.** A [`ControlScope::Subtree`]
//!    assignment reaches descendants; a [`ControlScope::DomainOnly`] one stays
//!    exactly where it was made. There is no scope that reaches a parent.
//! 2. **[`InheritanceMode::Isolated`] cuts the branch at itself.** An isolated
//!    domain does not read its parent's assignments, and neither does anything
//!    beneath it, because the walk from a descendant has to pass through the
//!    isolated node to get any higher.
//! 3. **Delegation never widens.** [`ControlIndex::insert_delegated`] refuses a
//!    role the delegator does not already cover, so no chain of delegations can
//!    manufacture a rung nobody held. That is the property the
//!    [`DomainError::AccessDomainControlWiden`] refusal exists to name.
//!
//! On top of those, a personal realm admits control for at most one principal.
//! ADR 0038 refuses membership on a personal project outright, and nesting
//! domains inside one must not become a quieter way to share: the moment a
//! second principal would gain control, [`ControlIndex::insert`] refuses with
//! [`DomainError::AccessDomainPersonalSharing`].

use super::forest::AccessDomainForest;
use crate::{AccessDomainId, DomainControlId, DomainError, PrincipalId};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// The three-rung ladder, mirroring the organization and project roles ADR 0038
/// already established rather than inventing a parallel vocabulary.
///
/// Declaration order *is* the ladder: `Ord` is derived from it, so
/// [`DomainRole::covers`] and the "strongest role wins" resolution both read
/// straight off the comparison instead of a hand-written table.
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DomainRole {
    Member,
    Admin,
    Owner,
}

impl DomainRole {
    /// Whether this rung is at least `other` — the check delegation uses.
    #[must_use]
    pub fn covers(self, other: Self) -> bool {
        self >= other
    }

    /// Whether this rung may rearrange domains and issue beneath them.
    ///
    /// A member holds a place in the estate, not authority over it.
    #[must_use]
    pub fn can_administer(self) -> bool {
        matches!(self, Self::Admin | Self::Owner)
    }

    /// Whether this rung may hand the domain itself to somebody else.
    #[must_use]
    pub fn can_transfer_ownership(self) -> bool {
        matches!(self, Self::Owner)
    }
}

/// How far an assignment reaches.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ControlScope {
    /// This domain and nothing else.
    DomainOnly,
    /// This domain and everything beneath it, until an isolated domain.
    Subtree,
}

/// Whether a domain reads control from above it.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum InheritanceMode {
    /// Inherit subtree assignments from ancestors.
    Inherit,
    /// Read only what was granted here. The branch below inherits nothing from
    /// above this node either.
    Isolated,
}

/// One principal's recorded control on one domain.
///
/// `Copy`, and identified by its own [`DomainControlId`], so an assignment can
/// be handed around, compared, and revoked by id without a caller having to
/// reconstruct the tuple it was made from.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ControlAssignment {
    pub id: DomainControlId,
    pub domain_id: AccessDomainId,
    pub principal_id: PrincipalId,
    pub role: DomainRole,
    pub scope: ControlScope,
}

/// The answer to "what does this principal hold *here*", after inheritance.
///
/// `source_domain_id` is the provenance, and it is the reason this is a struct
/// rather than a bare [`DomainRole`]: control that arrived by inheritance is
/// revoked at the domain it was granted on, so an operator shown an effective
/// role needs to be told where to go to remove it.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct EffectiveControl {
    pub principal_id: PrincipalId,
    /// The domain the question was asked about.
    pub domain_id: AccessDomainId,
    pub role: DomainRole,
    /// The domain the assignment actually lives on — equal to `domain_id` when
    /// the control was granted directly.
    pub source_domain_id: AccessDomainId,
}

/// Every control assignment in one realm's forest, keyed by assignment id.
///
/// Keyed by id rather than by domain so revocation is a single lookup and so
/// iteration order is stable: two indexes holding the same assignments resolve,
/// list and digest identically regardless of the order they were built in.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ControlIndex {
    assignments: BTreeMap<DomainControlId, ControlAssignment>,
}

impl ControlIndex {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    #[must_use]
    pub fn len(&self) -> usize {
        self.assignments.len()
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.assignments.is_empty()
    }

    /// Every assignment, in id order.
    pub fn assignments(&self) -> impl Iterator<Item = &ControlAssignment> {
        self.assignments.values()
    }

    /// The assignments recorded directly on `domain_id`.
    pub fn at(&self, domain_id: AccessDomainId) -> impl Iterator<Item = &ControlAssignment> + '_ {
        self.assignments
            .values()
            .filter(move |held| held.domain_id == domain_id)
    }

    /// Record an assignment.
    ///
    /// # Errors
    ///
    /// - [`DomainError::AccessDomainNotFound`] — the domain is not in `forest`,
    ///   which is also what keeps this index realm-bound: the forest refuses
    ///   foreign domains, so control cannot be recorded against one.
    /// - [`DomainError::AccessDomainConflict`] — the id is already recorded, or
    ///   this principal already holds control on this domain. One row per
    ///   principal per domain, so there is never a pair of rows whose combined
    ///   reading is stronger than either.
    /// - [`DomainError::AccessDomainPersonalSharing`] — a second principal would
    ///   gain control in a personal project.
    pub fn insert(
        &mut self,
        forest: &AccessDomainForest,
        assignment: ControlAssignment,
    ) -> Result<(), DomainError> {
        forest.require(assignment.domain_id)?;
        if self.assignments.contains_key(&assignment.id) {
            return Err(DomainError::AccessDomainConflict(format!(
                "control {} is already recorded",
                assignment.id
            )));
        }
        if self
            .at(assignment.domain_id)
            .any(|held| held.principal_id == assignment.principal_id)
        {
            return Err(DomainError::AccessDomainConflict(format!(
                "a principal already holds control on domain {}",
                assignment.domain_id
            )));
        }
        // Only a *second* principal is sharing. One person holding control on
        // several domains of their own estate is what a personal project is
        // for, so the refusal is keyed on the principal, not the row count.
        if self
            .assignments
            .values()
            .any(|held| held.principal_id != assignment.principal_id)
        {
            forest
                .realm()
                .assert_admits_sharing("a second controlling principal")?;
        }
        self.assignments.insert(assignment.id, assignment);
        Ok(())
    }

    /// Record an assignment made *by* `delegator`, refusing anything wider than
    /// what they hold.
    ///
    /// # Errors
    ///
    /// [`DomainError::AccessDomainControlWiden`] when the delegator holds no
    /// control on the domain, holds a role that does not administer it, or
    /// would hand out a rung above their own. Then everything
    /// [`ControlIndex::insert`] refuses — notably the personal-realm rule, so
    /// delegation is not a way around it.
    pub fn insert_delegated(
        &mut self,
        forest: &AccessDomainForest,
        delegator: PrincipalId,
        assignment: ControlAssignment,
    ) -> Result<(), DomainError> {
        let held = self
            .effective_control(forest, assignment.domain_id, delegator)?
            .ok_or_else(|| {
                DomainError::AccessDomainControlWiden(format!(
                    "{delegator} holds no control on domain {}",
                    assignment.domain_id
                ))
            })?;
        if !held.role.can_administer() {
            return Err(DomainError::AccessDomainControlWiden(format!(
                "{:?} does not administer domain {}",
                held.role, assignment.domain_id
            )));
        }
        if !held.role.covers(assignment.role) {
            return Err(DomainError::AccessDomainControlWiden(format!(
                "cannot delegate {:?} while holding {:?}",
                assignment.role, held.role
            )));
        }
        self.insert(forest, assignment)
    }

    /// Revoke one assignment by id.
    pub fn remove(&mut self, id: DomainControlId) -> Option<ControlAssignment> {
        self.assignments.remove(&id)
    }

    /// The strongest role granted *directly* on one rung of a walk.
    ///
    /// `inherited` says whether the rung is an ancestor rather than the domain
    /// the question was asked about. Domain-only control does not travel, so it
    /// is dropped on every rung but the first.
    fn strongest_at(
        &self,
        domain_id: AccessDomainId,
        principal_id: PrincipalId,
        inherited: bool,
    ) -> Option<DomainRole> {
        self.at(domain_id)
            .filter(|held| held.principal_id == principal_id)
            .filter(|held| !inherited || held.scope == ControlScope::Subtree)
            .map(|held| held.role)
            .max()
    }

    /// The strongest role `principal_id` holds on `domain_id`, after
    /// inheritance, or `None` when they hold nothing there.
    ///
    /// The walk goes from the domain upward and stops at the first isolated
    /// node, having counted what was granted on it. That ordering is what makes
    /// [`InheritanceMode::Isolated`] cover the whole branch rather than only the
    /// node it is set on: a descendant's walk cannot reach past it.
    ///
    /// # Errors
    ///
    /// [`DomainError::AccessDomainNotFound`] when the domain is absent, and
    /// [`DomainError::AccessDomainCycle`] when the parent walk revisits a node.
    pub fn effective_control(
        &self,
        forest: &AccessDomainForest,
        domain_id: AccessDomainId,
        principal_id: PrincipalId,
    ) -> Result<Option<EffectiveControl>, DomainError> {
        forest.require(domain_id)?;
        let mut chain = vec![domain_id];
        chain.extend(forest.ancestors(domain_id)?);
        let mut best: Option<EffectiveControl> = None;
        for (rung, source_domain_id) in chain.into_iter().enumerate() {
            let stronger = self
                .strongest_at(source_domain_id, principal_id, rung > 0)
                .filter(|role| best.is_none_or(|previous| *role > previous.role));
            if let Some(role) = stronger {
                best = Some(EffectiveControl {
                    principal_id,
                    domain_id,
                    role,
                    source_domain_id,
                });
            }
            if forest.require(source_domain_id)?.inheritance() == InheritanceMode::Isolated {
                break;
            }
        }
        Ok(best)
    }

    /// The administering control `principal_id` holds on `domain_id`.
    ///
    /// # Errors
    ///
    /// [`DomainError::AuthorizationDenied`] when the principal holds nothing
    /// there or holds a role that does not administer it, plus anything
    /// [`ControlIndex::effective_control`] reports.
    pub fn assert_may_administer(
        &self,
        forest: &AccessDomainForest,
        domain_id: AccessDomainId,
        principal_id: PrincipalId,
    ) -> Result<EffectiveControl, DomainError> {
        let held = self
            .effective_control(forest, domain_id, principal_id)?
            .ok_or_else(|| {
                DomainError::AuthorizationDenied(format!(
                    "{principal_id} holds no control on domain {domain_id}"
                ))
            })?;
        if !held.role.can_administer() {
            return Err(DomainError::AuthorizationDenied(format!(
                "{:?} does not administer domain {domain_id}",
                held.role
            )));
        }
        Ok(held)
    }
}
